File: VB/Wpf-Dashboard-Custom-Properties/Modules/ChartItemModule/ChartOptions/LineStyle.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Linq
Imports System.Windows.Data
Imports System.Windows.Media
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWpf
Imports DevExpress.Xpf.Charts
Imports DevExpress.XtraReports.UI

Namespace Wpf_Dashboard_Custom_Properties

    Public Class LineStyleConverter
        Implements System.Windows.Data.IMultiValueConverter

        Public Function Convert(ByVal values As Object(), ByVal targetType As System.Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements Global.System.Windows.Data.IMultiValueConverter.Convert
            Dim dashboardItemName As String = CStr(values(0))
            Dim dashboardControlProvider As IDashboardControlProvider = CType(values(1), IDashboardControlProvider)
            Dim valueDataMember As String = CStr(values(2))
            Dim dashStyle As System.Windows.Media.DashStyle = System.Windows.Media.DashStyles.Solid
            Dim chart As ChartDashboardItem = TryCast(dashboardControlProvider.Dashboard.Items(dashboardItemName), ChartDashboardItem)
            Dim series As SimpleSeries = chart.Panes.SelectMany(Function(p) p.Series).OfType(Of SimpleSeries)().FirstOrDefault(Function(s) s.Value.UniqueId Is valueDataMember)
            If series IsNot Nothing Then
                Dim propertyName As String = series.CustomProperties(Wpf_Dashboard_Custom_Properties.ChartItemModule.DashStylePropertyName)
                If Not System.[String].IsNullOrEmpty(propertyName) Then
                    Select Case propertyName.ToLower()
                        Case "dash"
                            dashStyle = System.Windows.Media.DashStyles.Dash
                        Case "dot"
                            dashStyle = System.Windows.Media.DashStyles.Dot
                        Case "dashdot"
                            dashStyle = System.Windows.Media.DashStyles.DashDot
                        Case "dashdotdot"
                            dashStyle = System.Windows.Media.DashStyles.DashDotDot
                    End Select
                End If
            End If

            Return New LineStyle(2) With {.DashStyle = dashStyle}
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As System.Type(), ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object() Implements Global.System.Windows.Data.IMultiValueConverter.ConvertBack
            Throw New System.NotImplementedException()
        End Function
    End Class

    Public Module LineStyleExtensions

        Public Sub CustomExport(ByVal dashboardControl As DashboardControl, ByVal e As CustomExportEventArgs)
            For Each keyValue As System.Collections.Generic.KeyValuePair(Of String, XRControl) In e.GetPrintableControls()
                If TypeOf keyValue.Value Is XRChart Then
                    Dim chartContext As ChartContext = e.GetChartContext(keyValue.Key)
                    Dim chartItem As ChartDashboardItem = TryCast(dashboardControl.Dashboard.Items(keyValue.Key), ChartDashboardItem)
                    Call Wpf_Dashboard_Custom_Properties.LineStyleExtensions.UpdateChart(chartItem, chartContext)
                End If
            Next
        End Sub

        Private Sub UpdateChart(ByVal chartItem As ChartDashboardItem, ByVal context As ChartContext)
            Dim lineSeries As System.Collections.Generic.List(Of SimpleSeries) = chartItem.Panes.SelectMany(Function(p) p.Series).OfType(Of SimpleSeries)().Where(Function(s) CBool((s.SeriesType Is SimpleSeriesType.Line))).ToList()
            For Each series As SimpleSeries In lineSeries
                Dim customPropertyValue As String = series.CustomProperties(Wpf_Dashboard_Custom_Properties.ChartItemModule.DashStylePropertyName)
                Dim dashStyle As DevExpress.XtraCharts.DashStyle
                If Not String.IsNullOrEmpty(customPropertyValue) AndAlso System.[Enum].TryParse(customPropertyValue, True, dashStyle) Then
                    For Each chartSeries In context.GetControlSeries(series)
                        Dim view As DevExpress.XtraCharts.LineSeriesView = TryCast(chartSeries.View, DevExpress.XtraCharts.LineSeriesView)
                        If view IsNot Nothing Then view.LineStyle.DashStyle = dashStyle
                    Next
                End If
            Next
        End Sub
    End Module
End Namespace

File: CS/Wpf-Dashboard-Custom-Properties/Modules/ChartItemModule/ChartOptions/LineStyle.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Windows.Data;
using System.Windows.Media;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWpf;
using DevExpress.Xpf.Charts;
using DevExpress.XtraReports.UI;

namespace Wpf_Dashboard_Custom_Properties {
    public class LineStyleConverter : IMultiValueConverter {
        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture) {
            string dashboardItemName = (string)values[0];
            IDashboardControlProvider dashboardControlProvider = (IDashboardControlProvider)values[1];
            string valueDataMember = (string)values[2];
            DashStyle dashStyle = DashStyles.Solid;
            ChartDashboardItem chart = dashboardControlProvider.Dashboard.Items[dashboardItemName] as ChartDashboardItem;
            SimpleSeries series = chart.Panes.SelectMany(p => p.Series).OfType<SimpleSeries>().FirstOrDefault(s => s.Value.UniqueId == valueDataMember);
            if(series != null) {
                string propertyName = series.CustomProperties[ChartItemModule.DashStylePropertyName];
                if(!String.IsNullOrEmpty(propertyName)) {
                    switch(propertyName.ToLower()) {
                        case "dash":
                            dashStyle = DashStyles.Dash;
                            break;
                        case "dot":
                            dashStyle = DashStyles.Dot;
                            break;
                        case "dashdot":
                            dashStyle = DashStyles.DashDot;
                            break;
                        case "dashdotdot":
                            dashStyle = DashStyles.DashDotDot;
                            break;
                    }
                }
            }
            return new LineStyle(2) { DashStyle = dashStyle };
        }
        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

    public static class LineStyleExtensions {
        public static void CustomExport(DashboardControl dashboardControl, CustomExportEventArgs e) {
            foreach(KeyValuePair<string, XRControl> keyValue in e.GetPrintableControls()) {
                if(keyValue.Value is XRChart) {
                    ChartContext chartContext = e.GetChartContext(keyValue.Key);
                    ChartDashboardItem chartItem = dashboardControl.Dashboard.Items[keyValue.Key] as ChartDashboardItem;
                    UpdateChart(chartItem, chartContext);
                }
            }
        }
        static void UpdateChart(ChartDashboardItem chartItem, ChartContext context) {
            List<SimpleSeries> lineSeries = chartItem.Panes.SelectMany(p => p.Series).OfType<SimpleSeries>().Where(s => s.SeriesType == SimpleSeriesType.Line).ToList();
            foreach(SimpleSeries series in lineSeries) {
                string customPropertyValue = series.CustomProperties[ChartItemModule.DashStylePropertyName];
                DevExpress.XtraCharts.DashStyle dashStyle;
                if(!string.IsNullOrEmpty(customPropertyValue) && Enum.TryParse(customPropertyValue, true, out dashStyle)) {
                    foreach(var chartSeries in context.GetControlSeries(series)) {
                        DevExpress.XtraCharts.LineSeriesView view = chartSeries.View as DevExpress.XtraCharts.LineSeriesView;
                        if(view != null)
                            view.LineStyle.DashStyle = dashStyle;
                    }
                }
            }
        }
    }
}
